core/trino-main/src/main/java/io/trino/execution/scheduler/StageEventDrivenTaskSourceFactory.java

```diff
@@ -0,0 +1,1310 @@
+            implements SplitAssigner
+    {
+        private final Optional<CatalogHandle> catalogRequirement;
+        private final Set<PlanNodeId> partitionedSources;
+        private final Set<PlanNodeId> replicatedSources;
+        private final Set<PlanNodeId> allSources;
+        private final long targetPartitionSizeInBytes;
+        private final Map<PlanNodeId, OutputDataSizeEstimate> outputDataSizeEstimates;
+        private final FaultTolerantPartitioningScheme sourcePartitioningScheme;
+        private final boolean preserveOutputPartitioning;
+
+        private Map<Integer, TaskPartition> outputPartitionToTaskPartition;
+        private final Set<Integer> createdTaskPartitions = new HashSet<>();
+        private final Set<PlanNodeId> completedSources = new HashSet<>();
+        private final ListMultimap<PlanNodeId, Split> replicatedSplits = ArrayListMultimap.create();
+
+        private int nextTaskPartitionId;
+
+        HashDistributionSplitAssigner(
+                Optional<CatalogHandle> catalogRequirement,
+                Set<PlanNodeId> partitionedSources,
+                Set<PlanNodeId> replicatedSources,
+                long targetPartitionSizeInBytes,
+                Map<PlanNodeId, OutputDataSizeEstimate> outputDataSizeEstimates,
+                FaultTolerantPartitioningScheme sourcePartitioningScheme,
+                boolean preserveOutputPartitioning)
+        {
+            this.catalogRequirement = requireNonNull(catalogRequirement, "catalogRequirement is null");
+            this.partitionedSources = ImmutableSet.copyOf(requireNonNull(partitionedSources, "partitionedSources is null"));
+            this.replicatedSources = ImmutableSet.copyOf(requireNonNull(replicatedSources, "replicatedSources is null"));
+            allSources = ImmutableSet.<PlanNodeId>builder()
+                    .addAll(partitionedSources)
+                    .addAll(replicatedSources)
+                    .build();
+            this.targetPartitionSizeInBytes = targetPartitionSizeInBytes;
+            this.outputDataSizeEstimates = ImmutableMap.copyOf(requireNonNull(outputDataSizeEstimates, "outputDataSizeEstimates is null"));
+            this.sourcePartitioningScheme = requireNonNull(sourcePartitioningScheme, "sourcePartitioningScheme is null");
+            this.preserveOutputPartitioning = preserveOutputPartitioning;
+        }
+
+        @Override
+        public AssignmentResult assign(PlanNodeId planNodeId, ListMultimap<Integer, Split> splits, boolean noMoreSplits)
+        {
+            AssignmentResult.Builder assignment = AssignmentResult.builder();
+            if (outputPartitionToTaskPartition == null) {
+                outputPartitionToTaskPartition = createOutputPartitionToTaskPartition();
+            }
+
+            if (replicatedSources.contains(planNodeId)) {
+                replicatedSplits.putAll(planNodeId, splits.values());
+                for (Integer partitionId : createdTaskPartitions) {
```

Comment:
I wanted to avoid unnecessary boxing unboxing. In this particular place it probably doesn't matter as only an unboxed `int` is needed in this loop, but in other loops there are callsites that are better off with a boxed Integer. I would prefer to keep it as is for consistency.